### PR TITLE
Audio Manager

### PR DESCRIPTION
**Pull request by keithliu03**
### Description
We now have a way to play audio/sfx.

### Trello Ticket
[Add the URL to the most applicable ticket on the [Trello](https://trello.com/b/fKcBhH3P/game-engine-development) board.](https://trello.com/c/05NsGYh5)

### Additional Info
Have not implemented audio manager into anything in the latest dev branch. Has been tested and does work with an outdated dev, and will implement once latest pull requests are done.

### CODE REVIEW

**Review by RaziqK — APPROVED** [review]

File: Code/Engine/AudioManager.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    public:
+        AudioManager(const AudioManager&) = delete;
+
+        static void loadAudioFile(const std::string& filename, const std::string& path, int volume, AudioType type);
+
+        static void playSFX(std::string filename);
+
+        static void playMusic(std::string filename, int times);
+
+        static void set_volume(std::string filename, int volume);
```
Comment (jake-pauls) [inline Code/Engine/AudioManager.h]: Style guide

- `UpperCamelCase` for method names
- `lowerCamelCase` for variable names (ie: `fileName`)

File: Code/Engine/AudioManager.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+void sad::AudioManager::loadAudioFile(const std::string& filename, const std::string& path, int volume, AudioType type)
+{
+    AudioFile audioFile;
+    audioFile.filename = filename;
+    audioFile.path = path.c_str();
+    audioFile.type = type;
+    audioFile.volume = volume;
+
+    core::Log(ELogType::Debug, "Audio Path: {}", path.c_str()); // Debug Path
```
Comment (jake-pauls) [inline Code/Engine/AudioManager.cpp]: Is this log necessary?

Prefer using `Info`, `Trace` or `Warn` logging channels if the logs being added are meant to be permanent/informational.

File: Code/Engine/AudioManager.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    if (Mix_OpenAudio(44100, MIX_DEFAULT_FORMAT, 2, 512) < 0) // Check if audio file can be opened
+    {
+        core::Log(ELogType::Debug, "Couldn't open audio sample: {}", Mix_GetError());
+    }
```
Comment (jake-pauls) [inline Code/Engine/AudioManager.cpp]: It looks like this is akin to SDL initialization code - we can probably move this check to `Window.cpp` as it only has to be done once and not for every audio file.

Namely, we should add `SDL_INIT_AUDIO` into the `SDL_Init` call on line 14 (recommended by the link below).
This method/check can then be added in the `sad::Window::Start` method with an assert.

Something like...
```
int mixerInitialization = Mix_OpenAudio(44100, MIX_DEFAULT_FORMAT, 2, 512);
SAD_ASSERT(mixerInitialization >= 0, "Failed to open audio device for SDL_Mixer");
```

https://wiki.libsdl.org/SDL_mixer/Mix_OpenAudio

File: Code/Engine/AudioManager.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                core::Log(ELogType::Debug, "Audio sample loaded!");
+                Mix_VolumeChunk(audioFile.mixChunk, volume);
+                
+                m_AudioFiles.push_back(audioFile);
+            }
+            else
+            {
+                core::Log(ELogType::Debug, "Couldn't load audio sample: {}", Mix_GetError());
+            }
+        }
```
Comment (jake-pauls) [inline Code/Engine/AudioManager.cpp]: We can probably make this super efficient and clean by leveraging the new ResourceManager.

I've already made an `AudioResource.cpp` that will have a handful of these settings automatically generated such as file name, path, and extension. What we can do is move the audio-file specific initialization code (such as mixing the volume chunks) to there and remove the static list of audio files in the audio manager.

In that sense, the `AudioManager` could just take a single `AudioResource` as most of the parameters (removing the static list and need for loops 🆒 ) here (ie for `SetVolume` or `PlaySFX`) and just play what's already been created for the audio resource by the resource manager.

Converted methods with `AudioResource` might look something like...
```
void sad::AudioManager::PlaySFX(const AudioResource& audioResource)
{
    if (Mix_Playing(-1) == 0)
    {
        Mix_PlayChannel(-1, audioResource.MixChunk, 0);
    }
} 
```

Playing an audio resource (in a script - when it comes 😄 ) or 'on the fly' would end up looking like this...
```
AudioResource* audioResource = ResourceManager::GetResource<AudioResource>("Code_Arabic_Nokia.wav");
AudioManager::GetInstance().PlaySFX(audioResource);
```

**Review by jake-pauls — CHANGES_REQUESTED** [review]
Awesome! Love that sound is in.

Left some comments for integrating this with the new `ResourceManager` features that will really cut this down. Make sure to pull latest to get the new `AudioResource`. Also make sure not to recommit the `PlayerController` as it has been renamed!

Otherwise looks fun!

**Comment by jake-pauls** [comment]
[Create a sample AudioManager that can play, stop, and pause audio with SDL](https://trello.com/c/05NsGYh5/27-create-a-sample-audiomanager-that-can-play-stop-and-pause-audio-with-sdl)

**Review by kiwon132445 — APPROVED** [review]

**Review by jake-pauls — APPROVED** [review]

**Pull request merged**